libraries/SoftwareSerial/src/SoftwareSerial.cpp

```diff
@@ -0,0 +1,432 @@
+        if (cmp_value >= UINT16_MAX) {
+          clock_rate = clock_rate / 2;
+          pre *= 2;
+        }
+      } while (cmp_value >= UINT16_MAX);
+      timer.setPrescaleFactor(pre);
+      timer.setOverflow(cmp_value);
+      timer.setCount(0);
+      timer.attachInterrupt(&handleInterrupt);
+      timer.resume();
+    } else {
+      timer.detachInterrupt();
+    }
+    cur_speed = speed;
+  }
+}
+
+// This function sets the current object as the "listening"
+// one and returns true if it replaces another
+bool SoftwareSerial::listen()
+{
+  if (_receivePin >= 0) {
+    // wait for any transmit to complete as we may change speed
+    while (active_out)
+      ;
+    if (active_listener) {
+      active_listener->stopListening();
+    }
+    rx_tick_cnt = 0;
+    rx_bit_cnt = -1;
+    setSpeed(_speed);
+    active_listener = this;
+    if (!_half_duplex) {
+      active_in = this;
+    }
+    return true;
+  }
+  return false;
+}
+
+// Stop listening. Returns true if we were actually listening.
+bool SoftwareSerial::stopListening()
+{
+  if (active_listener == this) {
+    // wait for any output to complete
+    while (active_out)
+      ;
+    if (_half_duplex) {
+      setRXTX(false);
+    }
+    active_listener = nullptr;
+    active_in = nullptr;
+    // turn off ints
+    setSpeed(0);
+    return true;
+  }
+  return false;
+}
+
+inline void SoftwareSerial::setTX()
+{
+  if (_inverse_logic) {
+    LL_GPIO_ResetOutputPin(_transmitPinPort, _transmitPinNumber);
+  } else {
+    LL_GPIO_SetOutputPin(_transmitPinPort, _transmitPinNumber);
+  }
+  pinMode(_transmitPin, OUTPUT);
+}
+
+inline void SoftwareSerial::setRX()
+{
+  if (_receivePin > 0) {
+    pinMode(_receivePin, _inverse_logic ? INPUT_PULLDOWN : INPUT_PULLUP); // pullup for normal logic!
+  }
+}
+
+inline void SoftwareSerial::setRXTX(bool input)
+{
+  if (_half_duplex) {
+    if (input) {
+      if (active_in != this) {
+        setRX();
+        rx_bit_cnt = -1;
+        rx_tick_cnt = 1;
+        active_in = this;
+      }
+    } else {
+      if (active_in == this) {
+        setTX();
+        active_in = nullptr;
+      }
+    }
+  }
+}
+
+inline void SoftwareSerial::send()
+{
+  if (--tx_tick_cnt < 0) {
+    if (tx_bit_cnt++ < 10) {
+      // send data (including start and stop bits)
+      if (tx_buffer & 1) {
+        LL_GPIO_SetOutputPin(_transmitPinPort, _transmitPinNumber);
+      } else {
+        LL_GPIO_ResetOutputPin(_transmitPinPort, _transmitPinNumber);
+
+      }
+      tx_buffer >>= 1;
+      tx_tick_cnt = OVERSAMPLE - 1; // first sample sent, OVERSAMPLE - 1 more samples to go
+    } else {
+      tx_tick_cnt = 0;
+      if (_output_pending || !(_half_duplex && active_listener == this)) {
+        active_out = nullptr;
+        // When in half-duplex mode, we wait for HALFDUPLEX_SWITCH_DELAY bit-periods after the byte has
+        // been transmitted before allowing the switch to RX mode
+      } else if (tx_bit_cnt > 10 + HALFDUPLEX_SWITCH_DELAY) {
```

Comment:
I think that if 5 bits are required you need (
* HALFDUPLEX_SWITCH_DELAY * OVERSAMPLE     (old style when send() check if (--tx_tick_cnt <= 0))
*  (HALFDUPLEX_SWITCH_DELAY * OVERSAMPLE) - 1  (new style with if (--tx_tick_cnt < 0) )